Math for Curves and Interpolations (Lerp and Callmul-Rom)

lassade opened the pull request:
Required by #1429, adds curves and a `Lerp` trait used by the animation system in the `bevy_math` crate;

Notes
- I didn't know what lerp algorithm was used by Vec2, Vec3 and Vec4 so I used one with precise t == 0 and t == 1
- `lerp` is by default clamped (but there is a `lerp_unclampled`)

Missing bits, (non block)
- [x] `Handle` and `HandleUntyped` should implement lerp using a step function
- [x] Test `CurveVariable` behaviour

The new `Color` definition make lerping and a color curve really complex and impractical (please join the discussion [here](https://github.com/bevyengine/bevy/discussions/1849)), It's is some what related but doesn't block this issue;

jamadazi commented:
Have you looked at the `bevy_easings` crate as prior art? Would anything from there be an inspiration / to be included into bevy?

Also, would your animations work obsolete `bevy_easings`? It would be great if the use cases that crate offers could be provided by bevy internally.

mockersf commented:
In the scope of #1429, this is much more general than `bevy_easings`.
For interpolation functions, they all at their core use the `lerp` method, so once it's available it gets trivial to add all the functions you want


There is commented code in this PR, should it be removed?

About the `TODO`s, do they need to be fixed in this PR or leaved for later?

Also, some of the doc comments don't sound completely right to my non-english native ear, could someone check?

lassade commented:
What comented code? I wan't to keep the `Color` stuff around because I know we are going to need them;

Not sure what TODOs, but I want to keep them for later, otherwise they will be forgoten;

Can you point out the comments that sound strange?

mockersf commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,123 @@
+        &mut self.keyframes[..]
+    }
+
+    // pub fn insert(&mut self, time_sample: f32, value: T) {
+    // }
+
+    // pub fn remove(&mut self, index: usize) {
+    //assert!(samples.len() > 1, "curve can't be empty");
+    // }
```
should this be kept?

mockersf commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,123 @@
+        (-self.negative_offset) as isize
+    }
+
+    // pub fn iter(&self) -> impl Iterator<Item = (f32, &T)> {
+    //     self.samples.iter().copied().zip(self.keyframes.iter())
+    // }
+
+    // pub fn iter_mut(&mut self) -> impl Iterator<Item = (f32, &mut T)> {
+    //     self.samples.iter().copied().zip(self.keyframes.iter_mut())
+    // }
```
and this?

mockersf commented on crates/bevy_math/src/curves/variable.rs:
```diff
@@ -0,0 +1,230 @@
+//             keyframes: values,
+//         }
+//     }
+
+//     pub fn from_line(t0: f32, t1: f32, v0: T, v1: T) -> Self {
+//         Self {
+//             time_stamps: if t1 >= t0 { vec![t0, t1] } else { vec![t1, t0] },
+//             keyframes: vec![v0, v1],
+//         }
+//     }
+
+//     pub fn from_constant(v: T) -> Self {
+//         Self {
+//             time_stamps: vec![0.0],
+//             keyframes: vec![v],
+//         }
+//     }
+
+//     // pub fn insert(&mut self, time_sample: f32, value: T) {
+//     // }
+
+//     // pub fn remove(&mut self, index: usize) {
+//     //assert!(samples.len() > 1, "curve can't be empty");
+//     // }
+
+//     pub fn add_offset_time(&mut self, time_offset: f32) {
+//         self.time_stamps.iter_mut().for_each(|t| *t += time_offset);
+//     }
+
+//     pub fn iter(&self) -> impl Iterator<Item = (f32, &T)> {
+//         self.time_stamps.iter().copied().zip(self.keyframes.iter())
+//     }
+
+//     pub fn iter_mut(&mut self) -> impl Iterator<Item = (f32, &mut T)> {
+//         self.time_stamps
+//             .iter()
+//             .copied()
+//             .zip(self.keyframes.iter_mut())
+//     }
+// }
```
and this?

lassade replied:
`CurveVariable` will need redesign

mockersf commented on crates/bevy_math/src/curves/variable_linear.rs:
```diff
@@ -0,0 +1,210 @@
+        }
+    }
+
+    // pub fn insert(&mut self, time_sample: f32, value: T) {
+    // }
+
+    // pub fn remove(&mut self, index: usize) {
+    //assert!(samples.len() > 1, "curve can't be empty");
+    // }
```
and this?

lassade replied:
I'm rounding up the public api, next commit they will be implemented

mockersf commented on crates/bevy_math/src/interpolation/interpolate.rs:
```diff
@@ -0,0 +1,156 @@
+    }
+}
+
+// impl Interpolate for Color {
+//     type Tangent = Self;
+
+//     fn interpolate(k0: &Self, k1: &Self, interp: &Interpolation<Self::Tangent>, t: f32) -> Self {
+//         match interp {
+//             Interpolation::Step => utils::step(k0, k1, t),
+//             Interpolation::Linear => utils::lerp(*k0, *k1, t),
+//             Interpolation::Smooth { right, left } => utils::catmull_rom::<Vec4>(
+//                 (*k0).into(),
+//                 (*right).into(),
+//                 (*k1).into(),
+//                 (*left).into(),
+//                 t,
+//             )
+//             .into(),
+//         }
+//     }
+// }
```
maybe a comment on this on why it's commented for color?

mockersf commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,123 @@
+/// Curve with evenly spaced keyframes, in another words a curve with a fixed frame rate.
+///
+/// This curve maintains the faster sampling rate over a wide range of frame rates, because
+/// it doesn't rely on keyframe cursor. But as a downside it will have the bigger memory foot print;
```
```suggestion
/// it doesn't rely on keyframe cursor. As a downside, it will have a bigger memory foot print.
```

mockersf commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,123 @@
+    // ? NOTE: Has I learned from benches casting to f32 is quite expensive
+    // ? so frame rate and offset values must be stored as f32
+    frame_rate: f32,
+    /// Negative number of frames before the curve starts
+    negative_offset: f32,
```
negative offset seems very strange to me while reading the code, could it be changed so that's it's a normal offset?

lassade replied:
It's an internal value meant, I do keep as negative value to improve perf, the public api handles the offset as it should be

mockersf commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,123 @@
+        self.keyframes.len()
+    }
+
+    /// True when doesn't have any keyframe
```
```suggestion
    /// `true` when this `CurveFixed` doesn't have any keyframe
```

mockersf commented on crates/bevy_math/src/curves/mod.rs:
```diff
@@ -0,0 +1,24 @@
+
+    fn duration(&self) -> f32;
+
+    /// Easer to use sampling method that don't have time restrictions or needs
+    /// the keyframe cursor, but is more expensive always `O(n)`. Which means
+    /// sampling takes longer to evaluate as much as time get closer to curve duration
+    /// and it get worse with more keyframes.
```
```suggestion
    /// Easier to use sampling method that doesn't have time restrictions or needs
    /// the keyframe cursor, but is more expensive (always `O(n)`). This means
    /// sampling takes longer to evaluate as time gets closer to curve duration.
    /// and it get worse with more keyframes.
```

mockersf commented on crates/bevy_math/src/curves/variable.rs:
```diff
@@ -0,0 +1,230 @@
+    }
+}
+
+// #[cfg(test)]
```
should this tests be commented?

lassade replied:
No actually I just redesigned the `CurveVariable` type, next commits I will be creating tests for it

mockersf commented on crates/bevy_math/src/curves/variable_linear.rs:
```diff
@@ -0,0 +1,210 @@
+// structured as SOA (struct of arrays), so the vec's length must match
+
+// TODO: impl Serialize, Deserialize
+/// Curve with sparse keyframes frames, in another words a curve with variable frame rate;
+///
+/// Very useful curve, because can accommodate the output of a linear reduction keyframe algorithm
+/// to lower the memory foot print. But as a down side require the use of keyframe cursor, and
+/// loses performance when the curve frame rate is higher than the curve sampling frame rate
```
```suggestion
/// Curve with sparse keyframes, in another words a curve with variable frame rate;
///
/// This is a very useful curve, because it can accommodate the output of a linear reduction keyframe algorithm
/// to lower the memory foot print. As a down side it requires the use of a keyframe cursor, and
/// loses performance when the curve frame rate is higher than the curve sampling frame rate.
```

mockersf commented on crates/bevy_math/src/interpolation/interpolate.rs:
```diff
@@ -0,0 +1,156 @@
+impl Interpolate for Quat {
+    type Tangent = Self;
+
+    /// Performs an nlerp, because it's much cheaper and easer to combine with other animations,
```
```suggestion
    /// Performs an nlerp, because it's cheaper and easier to combine with other animations,
```

mockersf commented on crates/bevy_math/src/interpolation/lerp.rs:
```diff
@@ -0,0 +1,99 @@
+
+    /// Lerp, `t` is clamped in [0; 1] range
+    fn lerp(a: &Self, b: &Self, t: f32) -> Self {
+        let t = t.min(1.0).max(0.0);
```
```suggestion
        let t = t.clamp(0.0, 1.0);
```
https://doc.rust-lang.org/std/primitive.f32.html#method.clamp

mockersf commented on crates/bevy_math/src/interpolation/lerp.rs:
```diff
@@ -0,0 +1,99 @@
+// }
+
+impl Lerp for Quat {
+    /// Performs an nlerp, because it's much cheaper and easer to combine with other animations,
```
```suggestion
    /// Performs an nlerp, because it's cheaper and easier to combine with other animations,
```

mockersf commented:
I added a few comments on commented code and 

> Not sure what TODOs, but I want to keep them for later, otherwise they will be forgoten;

There are 8 `TODO`s, will you address them in a next PR? Or need some help with them?

lassade commented:
I finished the `CurveVariable` redesign but I need some input on the API, I think the only thing left are the tests.

Bare in mind `CurveVariable` is the in equivalent to what Unity exposes to users to animate, so I plan to build an editor around it, that's why it also have the `TangentControl` associate with it (editor only value);

lassade commented:
Ok I think I got this done; I had rewrite the `CurveVariable` and tested using my on curve editor here [https://github.com/lassade/bevy_curve_editor](https://github.com/lassade/bevy_curve_editor), sady you it's quite hard to build because it depends on the egui to work since last update broke egui you will need to use the d5ad8401;

jakobhellermann commented:
> sadly you it's quite hard to build because it depends on the egui to work since last update broke egui

I created a [wip branch](https://github.com/mvlabat/bevy_egui/pull/25) of `bevy_egui` that works with bevy main:
```toml
bevy_egui = { git = "https://github.com/jakobhellermann/bevy_egui", branch = "bevy-main" }

[patch."https://github.com/bevyengine/bevy"]
bevy = { git = "https://github.com/lassade/bevy", branch = "curves-and-lerp" }
```

jakobhellermann commented on crates/bevy_math/src/interpolation/lerp.rs:
```diff
@@ -0,0 +1,109 @@
+
+impl<T: Lerp + Clone> Lerp for Option<T> {
+    fn lerp_unclamped(a: &Self, b: &Self, t: f32) -> Self {
+        match (a.is_some(), b.is_some()) {
+            (true, true) => Some(T::lerp_unclamped(
+                a.as_ref().unwrap(),
+                b.as_ref().unwrap(),
+                t,
+            )),
+            (false, true) | (true, false) | (false, false) => {
+                if t > 0.99 {
+                    b.clone()
+                } else {
+                    a.clone()
+                }
+            }
+        }
```
You can avoid the `unwrap`s if you match on the options directly:
```suggestion
        match (a, b) {
            (Some(a), Some(b)) => Some(T::lerp_unclamped(a, b, t)),
            _ => {
                if t > 0.99 {
                    b.clone()
                } else {
                    a.clone()
                }
            }
        }
```

msvbg commented on crates/bevy_math/src/curves/mod.rs:
```diff
@@ -0,0 +1,38 @@
+#[derive(Error, Debug)]
+pub enum CurveCreationError {
+    #[error("number of keyframes time stamps and values doesn't match")]
+    MissMachLength,
```
Almost!
```suggestion
    MismatchedLength,
```

ValorZard commented:
Any interest in merging this? It seems done and small, and easy to merge as well. Its also super important for any future animation stuff.

alice-i-cecile commented on crates/bevy_asset/src/handle.rs:
```diff
@@ -230,6 +231,34 @@ impl<T: Asset> Clone for Handle<T> {
 
+impl<T: Asset + 'static> Lerp for Handle<T> {
+    #[inline(always)]
+    fn lerp_unclamped(a: &Self, b: &Self, t: f32) -> Self {
```
I would like to see more descriptive variable names here. I can piece together a = "starting_position", b = "end_position", t = "time_step" from context but that will be harder for beginners and tough to read.

lassade replied:
from `a` to `b` is kind "standard" in the lerp functions, maybe docs should solve the issue?

alternatively I could agree with `v0` to `v1` meaning value at 0 and 1, what do you think?

alice-i-cecile replied:
I'm fine with `a` and `b` if they're standard, so long as they're documented in a way that's attached to the function :)

alice-i-cecile commented on crates/bevy_asset/src/handle.rs:
```diff
@@ -230,6 +231,34 @@ impl<T: Asset> Clone for Handle<T> {
     }
 }
 
+impl<T: Asset + 'static> Lerp for Handle<T> {
```
I'm a bit nervous about chucking Lerp and related methods onto *all* handles to assets. What if we end up using the asset system for audio? Or dialog?

Perhaps an extra trait + trait bound makes sense here, to limit this to assets with a transform.

lassade replied:
Handle fields can only be swapped, because `lerp` can't create a new asset handle to return, that's by desing.

What you are describing is done via `impl<T> Lerp for T` not by `impl<T> Lerp for Handle<T>`, even still I think lerping should only be done in copy types;

A new PR could add a trait that describes `fn lerp(&a, &b, t, &mut target)`

alice-i-cecile commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,121 @@
+/// it doesn't rely on keyframe cursor. As a downside, it will have a bigger memory foot print.
+#[derive(Default, Debug)]
+pub struct CurveFixed<T> {
+    frame_rate: f32,
```
I would like to see the units of the f32 documented in some way. Are these in seconds? Milliseconds?

alice-i-cecile replied:
Oh they seem to be in Hz! That was very surprising to me.

lassade replied:
"frame rate" is the same as "frames per second" there's isn't another definition for it that I'm aware of

alice-i-cecile commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,121 @@
+pub struct CurveFixed<T> {
+    frame_rate: f32,
+    /// Negative number of frames before the curve starts
+    negative_offset: f32,
```
The number of frames is an f32? That seems surprising to me.

alice-i-cecile replied:
And yet this value is passed in and out as an integer. Hmm...

lassade replied:
integer to float conversion and vise versa are hot instructions, I always try to put the data in the most convenient format possible for the sample function;

Plus it's possible to have fractional frame rates like the 29.97 that you find in some videos;

lassade replied:
Addressed in the new update

alice-i-cecile commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,121 @@
+}
+
+impl<T> CurveFixed<T> {
+    pub fn from_keyframes(frame_rate: f32, offset: isize, keyframes: Vec<T>) -> Self {
```
I would like some guidance on choosing the value of `offset` here. I have no idea where to begin and this is part of the public API.

alice-i-cecile replied:
Probably most fitting in `set_offset`?

lassade replied:
If I name it `frame_offset`, does make it more clearer?

lassade replied:
@NathanSWard here

alice-i-cecile replied:
`frame_offset` is much clearer to me :)

alice-i-cecile commented on crates/bevy_math/src/curves/mod.rs:
```diff
@@ -0,0 +1,38 @@
+    /// but is more expensive in some types of curve, been always `O(n)`.
+    ///
+    /// This means sampling is more expensive to evaluate as the `time` gets bigger;
+    fn sample(&self, time: f32) -> Self::Output;
```
This seems like one of the most important methods; I'd be happy to see some more basic documentation targeted towards beginners who have never animated anything before here.

lassade replied:
Why ideas on what I should write about?

The purpose of this function is to sample a curve at a given time, there isn't more to it.

alice-i-cecile replied:
Hmm. Perhaps this is better addressed at the module / book-level actually. I think that capturing enough information for beginners to make this truly useful requires an Animation 101, which doesn't belong here.

alice-i-cecile commented on crates/bevy_math/src/curves/mod.rs:
```diff
@@ -0,0 +1,38 @@
+pub use variable::*;
+pub use variable_linear::*;
+
+pub type CurveCursor = u16;
```
As someone new to animation, I have *no* idea what this does on first glance, especially if all I had access to was the docs.

NathanSWard replied:
Also, I'm not a fan of type aliases, I much prefer strong types.
So I would recommend
```rust
pub struct CurveCursor(u16);
```

lassade replied:
That make easy to change it to a `u32` for instance in case we need more keys, but I doubt

lassade replied:
I made some changed and renamed it to KeyframeIndex

lassade replied:
doc created

alice-i-cecile commented on crates/bevy_math/src/curves/variable.rs:
```diff
@@ -0,0 +1,708 @@
+/// Interpolation is based on this [article](http://archive.gamedev.net/archive/reference/articles/article1497.html),
+/// it's very similar to the implementation used by Unity, except that tangents doesn't have weighted mode;
+///
+/// **NOTE** Keyframes count is limited by the [`CurveCursor`] size.
```
```suggestion
/// **NOTE** The maximum number of keyframes is limited by the capacity of [`CurveCursor`] (a `u16`)
```

alice-i-cecile commented on crates/bevy_math/src/curves/variable.rs:
```diff
@@ -0,0 +1,708 @@
+    interpolation::{Interpolate, Interpolation},
+};
+
+/// Keyframe tangents control mode
```
This doc comment isn't very useful to users who don't know what tangent control does.

lassade replied:
done

lassade replied:
better comments

alice-i-cecile commented on crates/bevy_math/src/curves/variable.rs:
```diff
@@ -0,0 +1,708 @@
+            return Err(CurveError::KeyframeLimitReached(CurveCursor::MAX as usize));
+        }
+
+        // Make sure the
```
Sentence fragment <3

lassade replied:
done

alice-i-cecile commented on crates/bevy_math/src/curves/variable.rs:
```diff
@@ -0,0 +1,708 @@
+        }
+    }
+
+    pub fn remove(&mut self, index: CurveCursor) {
```
```suggestion
    /// Removes the keyframe at the index specified
    pub fn remove(&mut self, index: CurveCursor) {
```

alice-i-cecile commented on crates/bevy_math/src/curves/variable.rs:
```diff
@@ -0,0 +1,708 @@
+        self.tangents_in.remove(i);
+        self.tangents_out.remove(i);
+
+        if i < self.keyframes.len() {
```
```suggestion
       // Do not adjust tangents at ends of the curve
        if i < self.keyframes.len() {
```

lassade replied:
uploading a new doc version

alice-i-cecile commented on crates/bevy_math/src/curves/variable.rs:
```diff
@@ -0,0 +1,708 @@
+        self.adjust_tangents_with_neighbors(i);
+    }
+
+    pub fn set_time(&mut self, index: CurveCursor, time: f32) -> Option<CurveCursor> {
```
What does this function do? Sets the keyframe selected to occur at the given time?

alice-i-cecile replied:
And it looks like the other keyframes are adjusted so then the intervals are preserved?

lassade replied:
it moves a keyframe in the given index to a different point in time, then it update the tangents of the neighboring keyframes that are set to `Auto`;

Maybe I should document that

lassade replied:
done

alice-i-cecile commented on crates/bevy_math/src/curves/variable_linear.rs:
```diff
@@ -0,0 +1,243 @@
+            return Err(CurveError::KeyframeLimitReached(CurveCursor::MAX as usize));
+        }
+
+        // Make sure the
```
Another sentence fragment; seemingly duplicated.

lassade replied:
done

alice-i-cecile commented on crates/bevy_math/src/interpolation/interpolate.rs:
```diff
@@ -0,0 +1,190 @@
+
+    #[inline]
+    fn interpolate(
+        k0: &Self,
```
The variable names in these traits are hard to follow without any doc strings.

NathanSWard replied:
Also, I wouldn't be surprised if this triggers a `clippy` `too-many-arguments` warning. 
I would suggest grouping the parameters into structs.
```rust
fn foo(a: u32, b: u32, c: i32, d: i32);

// vs

struct AB {
    a: u32,
    b: u32,
}

struct CD {
    c: i32,
    d: i32,
}

fn foo(ab: AB, cd: CD);
```
(granted the names in this example are bad)

lassade replied:
that will only add more boiler plate, clippy is happy

lassade replied:
done

alice-i-cecile commented on crates/bevy_math/src/interpolation/lerp.rs:
```diff
@@ -0,0 +1,105 @@
+        match (a, b) {
+            (Some(a), Some(b)) => Some(T::lerp_unclamped(a, b, t)),
+            _ => {
+                if t > 0.99 {
```
Could you leave a comment explaining this line? I am very surprised by this magic number.

lassade replied:
Sure, this is just the step function used to change Some to None and vice versa

mockersf replied:
Lerping is usually between 0 and 1. This is pretty much the `t >= 1.0` of float (`t + ε >= 1.0` with `ε = 0.01`)

lassade replied:
I changed for a call to the step function

lassade replied:
I'm now just calling step function I also commented a bit

alice-i-cecile commented on crates/bevy_math/src/interpolation/utils.rs:
```diff
@@ -0,0 +1,67 @@
+///
+/// Took this one from piston: https://github.com/PistonDevelopers/skeletal_animation
+#[inline]
+pub fn inv_sqrt(x: f32) -> f32 {
```
```suggestion
pub fn fast_inv_sqrt(x: f32) -> f32 {
```
This should be called out as approximate IMO.

lassade replied:
done

alice-i-cecile commented on crates/bevy_math/src/interpolation/utils.rs:
```diff
@@ -0,0 +1,67 @@
+
+/// Quake 3 fast inverse sqrt
+///
+/// Took this one from piston: https://github.com/PistonDevelopers/skeletal_animation
```
```suggestion
/// Implementation borrowed from Piston under the MIT License: https://github.com/PistonDevelopers/skeletal_animation
```

alice-i-cecile commented on examples/animations/smooth_curves.rs:
```diff
@@ -0,0 +1,177 @@
+            })
+            .collect::<Vec<_>>(),
+    );
+    mesh.set_attribute(Mesh::ATTRIBUTE_NORMAL, values(DIVS, [0.0f32, 0.0, 1.0]));
```
I would like comments explaining each of the lines where we're setting mesh attributes; it's hard to tell why those specific values are needed and what the various numbers correspond to.

alice-i-cecile commented:
Alright, reviewed! There a few nits, some inadequate documentation / variable names and a few to-dos that I think we might want to address before merging.

Overall I'm quite happy with the code quality and utility of this PR though. Good work!

alice-i-cecile commented on crates/bevy_math/src/interpolation/utils.rs:
```diff
@@ -0,0 +1,67 @@
+use std::ops::{Add, Div, Mul, Sub};
+
+/// Quake 3 fast inverse sqrt
```
I would be in favor of replacing this algorithm with the algorithm from [Walcyzk et al. 2018](https://arxiv.org/abs/1802.06302). It's claimed to be significantly more accurate without additional computation costs. Page 15 has the final algorithm that we would be implementing.

lassade replied:
I added a TODO note to do this later, make in the next revision

NathanSWard commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,121 @@
+    pub keyframes: Vec<T>,
+}
+
+impl<T: Clone> Clone for CurveFixed<T> {
```
Is it not possible to just `#[derive(Clone)]` on the original struct?

lassade replied:
Apparently yes!

NathanSWard commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,121 @@
+        ((self.keyframes.len() as f32 - 1.0 - self.negative_offset) / self.frame_rate).max(0.0)
+    }
+
+    fn sample(&self, time: f32) -> Self::Output {
```
This should be documented that it'll `panic!` if `self.keyframes` is empty.

lassade replied:
Is there any standard way of doing this?

lassade replied:
done, I'm writing following the the std lib as a guide

lassade replied:
btw, I added the panic docs in the trait not the impl

NathanSWard commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,121 @@
+    pub fn from_keyframes(frame_rate: f32, offset: isize, keyframes: Vec<T>) -> Self {
+        Self {
+            frame_rate,
+            negative_offset: -(offset as f32),
```
What's the purpose of having `offset` as an `isize` and not a `f32`?
The conversion to and from `f32` and not the best imo.

lassade replied:
I had a reason but in a second though maybe I should just use a `f32`

lassade replied:
I remember negative_offset should be called negative_frame_offset it's an integer stored in the `f32` for performance reasons

lassade replied:
there's a similar issue above, closing this one

NathanSWard commented on crates/bevy_math/src/curves/fixed.rs:
```diff
@@ -0,0 +1,121 @@
+
+    /// Same as `sample` function
+    #[inline]
+    fn sample_with_cursor(&self, cursor: CurveCursor, time: f32) -> (CurveCursor, Self::Output) {
+        let _ = cursor;
```
I you have an unused parameter, it's usually more idiomatic to prefix with an `_` underscore.
```suggestion
    fn sample_with_cursor(&self, _cursor: CurveCursor, time: f32) -> (CurveCursor, Self::Output) {
```

lassade replied:
Ok

NathanSWard commented on crates/bevy_math/src/curves/variable_linear.rs:
```diff
@@ -0,0 +1,243 @@
+
+    pub fn insert(&mut self, time: f32, value: T) {
+        // Keyframe length is limited by the cursor size yype that is 2 bytes,
+        assert!(
```
ditto on documenting the `panic!` behavior.

lassade replied:
done

NathanSWard commented on crates/bevy_math/src/curves/variable_linear.rs:
```diff
@@ -0,0 +1,243 @@
+/// **NOTE** Keyframes count is limited by the [`CurveCursor`] size.
+#[derive(Default, Debug)]
+pub struct CurveVariableLinear<T> {
+    time_stamps: Vec<f32>,
+    keyframes: Vec<T>,
```
Is there a specific reason to not use `Vec<(f32, T)>`?

lassade replied:
Yes, TLDR: Is usually faster.

The `sample` function perform a search to find the right keyframes for any given sample time, this search can be made more efficient by not pouting the cache with unused values like T in this case. This translates to faster random time sampling, but it also makes continuous sampling a bit faster

lassade commented:
Added 2 more functions in the Curve api to move common functions from curve specific impl to the trait impl;
Added a resampling function that I wrote a while back

NathanSWard commented on crates/bevy_math/src/interpolation/lerp.rs:
```diff
@@ -0,0 +1,99 @@
+/// Defines how a particular type will be interpolated
+pub trait Lerp: Sized {
+    /// Lerp, `t` is unclamped
+    fn lerp_unclamped(a: &Self, b: &Self, t: f32) -> Self;
```
I don't think `t` should be bound to an `f32`.
Imo this should be an associated type.

lassade replied:
t is the lerping factor that ranges from 0 to 1

NathanSWard replied:
Yes but it's possible a user could want to use an `f64` or someother fixed point type for example.

alice-i-cecile commented on crates/bevy_math/src/interpolation/utils.rs:
```diff
@@ -0,0 +1,78 @@
+///
+/// Implementation borrowed from Piston under the MIT License: [https://github.com/PistonDevelopers/skeletal_animation]
+#[inline]
+pub fn fast_inv_sqrt(x: f32) -> f32 {
```
Alright, I did some more digging: Rust's native solution for this is faster than the approximate solution because inverse square root has a dedicated instruction these days. See: https://stackoverflow.com/a/59082210/3234149 

IMO this function should be cut completely and its usages should be replaced with `sqrt().recip()`.

lassade replied:
Good catch

lassade replied:
Do you know how the intrinsic rsqrt + a single Newton-Raphson iteration compares with the 'carmack' method?

In that same thread I found this: https://stackoverflow.com/questions/31555260, so I compiled a few possible methods here https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=ddc10be842400c95b81a0c5fba218da6;

Maybe my memory is failing me but a old bench showed me that `sqrtss` as a very hot instruction and that made me use  the carmak method, I will have to double check that ...

james7132 replied:
Small note: While `sqrt` and `1.0/x` are guaranteed under IEEE-754 2008, implementations at the hardware level may or may not be compliant. My particular game I'm working on requires animation to be in the networking simulation hotpath, and cross-platform determinism is mandatory. If we are going down the route of `sqrt().recip()`, could we use `libm::sqrtf` here under a feature flag? glam already has a feature flag for it, so we could pass the feature flag through for ease of use.

The 'carmack' method here does not have this issue and should be consistent across platforms.

alice-i-cecile replied:
Yeah, I think we're going to need to do our own modern benchmarks here to get a firm answer on what's correct for this context.

The note on determinism is important; the approach James suggests seems very reasonable.

lassade replied:
My benches gave-me these results:

| function | time | maximum relative error |
| ---------| ----- | ----------------------- |
| fpu (sqrt + div) | 1.4637 ns | 0% |
| quake3 | 1.1379 ns | 0.17507957% |
| _mm_rsqrt_ss | 846.43 ps | 0.029162358% |
| _mm_rsqrt_ss + 1nr iteration | 1.0436 ns | 0.000020642221%  |

alice-i-cecile replied:
Awesome; those are good numbers to have. How hard would it be to ensure input alignment?

I'm in favor of more accuracy + smaller, more legible code base over the micro-optimization here, but I'll defer to others' judgement in the final review.

james7132 replied:
Directly using `_mm_rsqrt_ss` may affect portability, we'll need to find alternatives for non-x86 platforms (i.e. Android/iOS). We'll probably need some SIMD abstraction for this like `simba`. Easiest way to ensure alignment in Rust is to use a `#[repr(C, align(16)]` types like glam's Vec3A/Vec4. 

Given these results, I'm actually more in favor for just using the fast_inv_sqrt as is since it's guaranteed to be bit-for-bit identical across targets, it's a notable speedup over the normal `sqrt().recip()`, and the error is still within tolerable ranges for a lot of the target usecases for such a function.

lassade replied:
I think in doing `_mm_rsqrt_ss ` with a fallback to `quake3` method for non `x86` or `libm` / `no_std` but I'm not sure if the aligment requiriment will be a problem, expecially when called by the consumer code, on the nlerp side I can make things rigth I think ...

lassade replied:
@james7132 did you checked the lib impl for sqrtf [here](https://github.com/rust-lang/libm/blob/master/src/math/sqrtf.rs)? It don't look like it's cross-plataform deterministic

lassade replied:
I submitted a revision for this function, by the looks of it the compiler can optimize the re-alignment instructions so it should not be a problem

james7132 replied:
It doesn't look like it at a first glance and it's difficult to tell if it actually produces the same output as the intrinsic with `sse` disabled. If it does, we may need to reevaluate this approach to achieving cross-platform FP determinism.

The `approx_rsqrt` implementation here, however, we for a fact know does not produce the same bit-for-bit results between the two implementations given the error difference seen from the benchmarks.

lassade replied:
I suggest you to resample the curve animations, that way you can mitigate any differences between cross plataform.

Weibye reviewed (changes requested):
The `examples/animations/smooth_curves.rs` needs an entry in the `examples/README.md`

alice-i-cecile commented:
@lassade, can you rebase this?

@james7132 is this still useful for your animation plans?

alice-i-cecile commented:
@lassade, can you rebase this?

@james7132 is this still useful for your animation plans?

james7132 commented:
The curves are definitely still needed. However, I think it would be better to use the `Animatable` trait over `Lerp` as it also addresses blending values a more clean way.

lassade commented:
> @lassade, can you rebase this?

I can't  :(

alice-i-cecile commented:
Fair enough. I'll close this out; we should reuse (and credit) this code later when attempting to tackle this down the road.
